lib/fastlane/plugin/wpmreleasetoolkit/actions/common/buildkite_upload_pipeline_action.rb

```diff
@@ -0,0 +1,74 @@
+module Fastlane
+  module Actions
+    class BuildkiteUploadPipelineAction < Action
+      def self.run(params)
+        env_file = params[:env_file]
+        pipeline_file = params[:pipeline_file]
+        branch = params[:branch]
+        commit = params[:commit]
+
+        UI.user_error!("Pipeline file not found: #{pipeline_file}") unless File.exist?(pipeline_file)
+        UI.user_error!('You should not provide both `branch` and `commit`') if !branch.nil? && commit != 'HEAD'
+        UI.user_error!('This action can only be called from a Buildkite CI build') unless ENV['BUILDKITE'] == 'true'
+
+        UI.message "Uploading pipeline on #{pipeline_file}, #{"branch #{branch}, " if branch}commit #{commit}"
+
+        env_vars = {
+          'BUILDKITE_BRANCH' => branch,
+          'BUILDKITE_COMMIT' => commit
+        }.compact
+
+        if env_file && File.exist?(env_file)
+          UI.message(" - Sourcing environment file beforehand: #{env_file}")
+
+          sh(env_vars, ". #{env_file.shellescape} && buildkite-agent pipeline upload #{pipeline_file.shellescape}")
```

Comment:
Nitpick: I think that using `source` instead of `.` is more readable.

 - I know that `.` is technically more portable as it's POSIX-standard, but by now most shells (at least bash and zsh) have `source` too, so that shouldn't be a problem
 - [We use `source` and not `.` in the root steps of our pipelines in our `*.tf` files too](https://github.com/Automattic/buildkite-ci/blob/trunk/src/buildkite/pipelines/_pipeline_upload_steps.tftpl.yml#L4) so we might as well be consistent, given one of the goals of this `builkite_upload_pipeline` action is to do the same steps (sourcing a bash file before doing pipeline upload) as when we upload a pipeline at the very start of a build 😉

Reply:
I switched to `.` while trying out alternatives for running `sh`, but totally agree with the point with consistency and with it being more familiar / easier to read. Updated on b062ef94afd01722f01dafee15b6567b969bc48b.